File: backend/app/services/package_search_service.py
"""Staff package lookup by Package Boss ID or carrier tracking (partial match)."""

from __future__ import annotations

from sqlalchemy import or_

from app.models.package import Package
from app.services.pre_alert_service import tracking_core

MIN_PACKAGE_SEARCH_QUERY_LEN = 5
PREFILTER_LIMIT = 250
DEFAULT_RESULT_LIMIT = 20

PB_EXACT_SCORE = 20_000
PB_PREFIX_BASE = 15_000
CARRIER_EXACT_BASE = 10_000


def _carrier_search_score(query_core: str, carrier: str | None) -> int:
    carrier_core = tracking_core(carrier)
    if not query_core or not carrier_core:
        return 0

    if query_core == carrier_core:
        return CARRIER_EXACT_BASE + len(query_core)

    short, long = (
        (query_core, carrier_core)
        if len(query_core) <= len(carrier_core)
        else (carrier_core, query_core)
    )
    if len(short) < MIN_PACKAGE_SEARCH_QUERY_LEN:
        return 0
    if short not in long:
        return 0

    return len(short)


def _match_type(score: int, field: str) -> str:
    if field == "tracking_number":
        return "exact" if score >= PB_EXACT_SCORE else "prefix"
    if score >= CARRIER_EXACT_BASE:
        return "exact"
    return "partial"


def search_packages(query: str, *, limit: int = DEFAULT_RESULT_LIMIT) -> tuple[list[dict], bool]:
    q = (query or "").strip()
    if len(q) < MIN_PACKAGE_SEARCH_QUERY_LEN:
        return [], False

    limit = max(1, min(limit, 30))
    q_upper = q.upper()
    q_core = tracking_core(q)

    scored: dict[str, tuple[Package, int, str, str]] = {}
    truncated = False

    def add_match(pkg: Package, score: int, field: str, matched_value: str) -> None:
        pid = str(pkg.id)
        existing = scored.get(pid)
        if existing is None or score > existing[1]:
            scored[pid] = (pkg, score, field, matched_value)

    # Exact Package Boss tracking number
    exact_pkg = Package.query.filter_by(tracking_number=q_upper).first()
    if exact_pkg:
        add_match(exact_pkg, PB_EXACT_SCORE + len(q_upper), "tracking_number", exact_pkg.tracking_number)
        return [_format_match(*scored[str(exact_pkg.id)])], False

    # Prefix on PB-… (e.g. PB-2026-000)
    if q_upper.startswith("PB"):
        pb_rows = (
            Package.query.filter(Package.tracking_number.ilike(f"{q_upper}%"))
            .order_by(Package.received_at.desc(), Package.tracking_number.desc())
            .limit(PREFILTER_LIMIT + 1)
            .all()
        )
        if len(pb_rows) > PREFILTER_LIMIT:
            truncated = True
            pb_rows = pb_rows[:PREFILTER_LIMIT]
        for pkg in pb_rows:
            add_match(
                pkg,
                PB_PREFIX_BASE + len(q_upper),
                "tracking_number",
                pkg.tracking_number,
            )

    # Carrier tracking — ILIKE prefilter, then score with normalized core
    carrier_filters = [Package.carrier_tracking.isnot(None)]
    carrier_clauses = []
    if q_core and len(q_core) >= MIN_PACKAGE_SEARCH_QUERY_LEN:
        carrier_clauses.append(Package.carrier_tracking.ilike(f"%{q_core}%"))
    if q != q_core:
        carrier_clauses.append(Package.carrier_tracking.ilike(f"%{q}%"))
    if carrier_clauses:
        carrier_rows = (
            Package.query.filter(*carrier_filters)
            .filter(or_(*carrier_clauses))
            .order_by(Package.received_at.desc(), Package.tracking_number.desc())
            .limit(PREFILTER_LIMIT + 1)
            .all()
        )
        if len(carrier_rows) > PREFILTER_LIMIT:
            truncated = True
            carrier_rows = carrier_rows[:PREFILTER_LIMIT]
        for pkg in carrier_rows:
            score = _carrier_search_score(q_core or q_upper, pkg.carrier_tracking)
            if score > 0:
                add_match(
                    pkg,
                    score,
                    "carrier_tracking",
                    pkg.carrier_tracking or "",
                )

    ranked = sorted(
        scored.values(),
        key=lambda item: (-item[1], -item[0].received_at.timestamp(), item[0].tracking_number),
    )
    return [_format_match(pkg, score, field, matched) for pkg, score, field, matched in ranked[:limit]], truncated


def _format_match(pkg: Package, score: int, field: str, matched_value: str) -> dict:
    from app.services.package_service import warehouse_package_to_dict

    return {
        "package": warehouse_package_to_dict(pkg),
        "match_score": score,
        "match_field": field,
        "match_type": _match_type(score, field),
        "matched_value": matched_value,
    }

File: server/src/services/packageSearchService.ts
import { and, desc, eq, ilike, isNotNull, or } from 'drizzle-orm'
import { packages } from '../db/schema/index.js'

type PackageRow = typeof packages.$inferSelect
import { db } from '../db/index.js'
import { warehousePackageToDict } from '../lib/serializers/package.js'
import { trackingCore } from './preAlertService.js'

export const MIN_PACKAGE_SEARCH_QUERY_LEN = 5
const PREFILTER_LIMIT = 250
const DEFAULT_RESULT_LIMIT = 20

const PB_EXACT_SCORE = 20_000
const PB_PREFIX_BASE = 15_000
const CARRIER_EXACT_BASE = 10_000

function carrierSearchScore(queryCore: string, carrier: string | null | undefined): number {
  const carrierCore = trackingCore(carrier)
  if (!queryCore || !carrierCore) return 0
  if (queryCore === carrierCore) return CARRIER_EXACT_BASE + queryCore.length

  const [short, long] =
    queryCore.length <= carrierCore.length ? [queryCore, carrierCore] : [carrierCore, queryCore]
  if (short.length < MIN_PACKAGE_SEARCH_QUERY_LEN) return 0
  if (!long.includes(short)) return 0
  return short.length
}

function matchType(score: number, field: string): string {
  if (field === 'tracking_number') return score >= PB_EXACT_SCORE ? 'exact' : 'prefix'
  if (score >= CARRIER_EXACT_BASE) return 'exact'
  return 'partial'
}

function formatMatch(
  pkg: PackageRow,
  score: number,
  field: string,
  matchedValue: string,
): Record<string, unknown> {
  return {
    package: warehousePackageToDict(pkg),
    match_score: score,
    match_field: field,
    match_type: matchType(score, field),
    matched_value: matchedValue,
  }
}

export async function searchPackages(
  query: string,
  opts: { limit?: number } = {},
): Promise<[Array<Record<string, unknown>>, boolean]> {
  const q = (query ?? '').trim()
  if (q.length < MIN_PACKAGE_SEARCH_QUERY_LEN) return [[], false]

  const limit = Math.max(1, Math.min(opts.limit ?? DEFAULT_RESULT_LIMIT, 30))
  const qUpper = q.toUpperCase()
  const qCore = trackingCore(q)

  const scored = new Map<string, [PackageRow, number, string, string]>()
  let truncated = false

  function addMatch(pkg: PackageRow, score: number, field: string, matchedValue: string): void {
    const existing = scored.get(pkg.id)
    if (!existing || score > existing[1]) {
      scored.set(pkg.id, [pkg, score, field, matchedValue])
    }
  }

  const [exactPkg] = await db
    .select()
    .from(packages)
    .where(eq(packages.trackingNumber, qUpper))
    .limit(1)
  if (exactPkg) {
    addMatch(exactPkg, PB_EXACT_SCORE + qUpper.length, 'tracking_number', exactPkg.trackingNumber)
    const match = scored.get(exactPkg.id)!
    return [[formatMatch(...match)], false]
  }

  if (qUpper.startsWith('PB')) {
    const pbRows = await db
      .select()
      .from(packages)
      .where(ilike(packages.trackingNumber, `${qUpper}%`))
      .orderBy(desc(packages.receivedAt), desc(packages.trackingNumber))
      .limit(PREFILTER_LIMIT + 1)
    if (pbRows.length > PREFILTER_LIMIT) {
      truncated = true
      pbRows.length = PREFILTER_LIMIT
    }
    for (const pkg of pbRows) {
      addMatch(pkg, PB_PREFIX_BASE + qUpper.length, 'tracking_number', pkg.trackingNumber)
    }
  }

  const carrierClauses = []
  if (qCore && qCore.length >= MIN_PACKAGE_SEARCH_QUERY_LEN) {
    carrierClauses.push(ilike(packages.carrierTracking, `%${qCore}%`))
  }
  if (q !== qCore) {
    carrierClauses.push(ilike(packages.carrierTracking, `%${q}%`))
  }
  if (carrierClauses.length) {
    const carrierRows = await db
      .select()
      .from(packages)
      .where(and(isNotNull(packages.carrierTracking), or(...carrierClauses)))
      .orderBy(desc(packages.receivedAt), desc(packages.trackingNumber))
      .limit(PREFILTER_LIMIT + 1)
    if (carrierRows.length > PREFILTER_LIMIT) {
      truncated = true
      carrierRows.length = PREFILTER_LIMIT
    }
    for (const pkg of carrierRows) {
      const score = carrierSearchScore(qCore || qUpper, pkg.carrierTracking)
      if (score > 0) {
        addMatch(pkg, score, 'carrier_tracking', pkg.carrierTracking ?? '')
      }
    }
  }

  const ranked = [...scored.values()].sort((a, b) => {
    if (b[1] !== a[1]) return b[1] - a[1]
    const aTime = a[0].receivedAt?.getTime() ?? 0
    const bTime = b[0].receivedAt?.getTime() ?? 0
    if (bTime !== aTime) return bTime - aTime
    return a[0].trackingNumber.localeCompare(b[0].trackingNumber)
  })

  return [ranked.slice(0, limit).map(([pkg, score, field, matched]) => formatMatch(pkg, score, field, matched)), truncated]
}
